File: apps/api/src/schemas/tools.py
"""Tool / version API shapes (M3g + M5c draft state)."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator


class CamelModel(BaseModel):
    model_config = ConfigDict(
        populate_by_name=True,
        serialize_by_alias=True,
        extra="ignore",
    )


class ToolVersionResponse(CamelModel):
    id: str
    tool_id: str = Field(alias="toolId")
    target: str
    code: str
    param_schema: Any = Field(alias="paramSchema")
    default_params: Any = Field(alias="defaultParams")
    asset_slots: Any = Field(alias="assetSlots")
    plan: Any | None = None
    created_at: str = Field(alias="createdAt")


class ToolResponse(CamelModel):
    id: str
    public_id: str = Field(alias="publicId")
    owner_user_id: str = Field(alias="ownerUserId")
    status: str
    title: str | None = None
    description: str | None = None
    tags: list[str] = Field(default_factory=list)
    published_at: str | None = Field(default=None, alias="publishedAt")
    published_version_id: str | None = Field(
        default=None,
        alias="publishedVersionId",
    )
    gallery_ready: bool = Field(default=False, alias="galleryReady")
    export_smoke_at: str | None = Field(default=None, alias="exportSmokeAt")
    thumbnail_asset_id: str | None = Field(
        default=None,
        alias="thumbnailAssetId",
    )
    thumbnail_url: str | None = Field(default=None, alias="thumbnailUrl")
    created_at: str = Field(alias="createdAt")
    updated_at: str = Field(alias="updatedAt")
    latest_version: ToolVersionResponse | None = Field(
        default=None,
        alias="latestVersion",
    )
    # M5c: Studio personalization (owner-only GET)
    draft_params: dict[str, Any] = Field(
        default_factory=dict,
        alias="draftParams",
    )
    draft_assets: dict[str, Any] = Field(
        default_factory=dict,
        alias="draftAssets",
    )
    # Continuous Studio refine transcript (tool-scoped)
    chat_history: list[Any] = Field(
        default_factory=list,
        alias="chatHistory",
    )


class ToolPublishRequest(CamelModel):
    """
    M8a/M8b optional publish body. All fields optional for thin M7 share compat.
    Empty body / omitted fields keep existing tool metadata.

    forGallery=true runs quality gates and sets galleryReady on success.
    exportSmokeOk must be true for gallery (client proved captureFrame/PNG).
    """

    title: str | None = None
    description: str | None = None
    tags: list[str] | None = None
    freeze_draft: bool = Field(default=False, alias="freezeDraft")
    for_gallery: bool = Field(default=False, alias="forGallery")
    export_smoke_ok: bool = Field(default=False, alias="exportSmokeOk")
    # M8c: asset id from POST /assets kind=thumb (omit to keep existing)
    thumbnail_asset_id: str | None = Field(
        default=None,
        alias="thumbnailAssetId",
    )


class PublishGateFailureItem(CamelModel):
    code: str
    message: str


class PublishGatesFailedDetail(CamelModel):
    """Structured 422 detail when gallery gates fail."""

    message: str = "Publish gates failed"
    code: str = "GATES_FAILED"
    gates: list[PublishGateFailureItem] = Field(default_factory=list)


class OwnerToolCardResponse(CamelModel):
    """Thin library card — no code, draft bags, or chat."""

    id: str
    public_id: str = Field(alias="publicId")
    title: str | None = None
    status: str
    gallery_ready: bool = Field(default=False, alias="galleryReady")
    thumbnail_url: str | None = Field(default=None, alias="thumbnailUrl")
    updated_at: str = Field(alias="updatedAt")
    published_at: str | None = Field(default=None, alias="publishedAt")
    is_remix: bool = Field(alias="isRemix")
    has_runnable_version: bool = Field(alias="hasRunnableVersion")
    tags: list[str] = Field(default_factory=list)


class OwnerToolListResponse(CamelModel):
    items: list[OwnerToolCardResponse]
    limit: int
    offset: int
    has_more: bool = Field(alias="hasMore")


class ToolDraftPatchRequest(CamelModel):
    """
    Full-replace bags for fields that are present.
    At least one of draftParams / draftAssets required.
    """

    draft_params: dict[str, Any] | None = Field(default=None, alias="draftParams")
    draft_assets: dict[str, Any] | None = Field(default=None, alias="draftAssets")

    @model_validator(mode="after")
    def _require_one_bag(self) -> ToolDraftPatchRequest:
        if self.draft_params is None and self.draft_assets is None:
            raise ValueError("Provide draftParams and/or draftAssets")
        return self


# ---------------------------------------------------------------------------
# M7d — public tool surface (anonymous read of published tools)
# ---------------------------------------------------------------------------


class PublicToolVersionResponse(CamelModel):
    """
    Version payload enough to mount in the public sandbox.
    Code is included for runtime load — not a download endpoint.
    """

    id: str
    target: str
    code: str
    param_schema: Any = Field(alias="paramSchema")
    default_params: Any = Field(alias="defaultParams")
    asset_slots: Any = Field(alias="assetSlots")


class PublicToolResponse(CamelModel):
    """
    Anonymous public tool view.
    No ownerUserId, no draftParams/draftAssets, no plan (owner/codegen noise).
    """

    public_id: str = Field(alias="publicId")
    status: str
    title: str | None = None
    description: str | None = None
    tags: list[str] = Field(default_factory=list)
    published_at: str | None = Field(default=None, alias="publishedAt")
    published_version_id: str | None = Field(
        default=None,
        alias="publishedVersionId",
    )
    thumbnail_asset_id: str | None = Field(
        default=None,
        alias="thumbnailAssetId",
    )
    thumbnail_url: str | None = Field(default=None, alias="thumbnailUrl")
    version: PublicToolVersionResponse

File: apps/web/lib/api/tools.ts
import { getApiBaseUrl } from "./config";

export type ToolVersionResponse = {
  id: string;
  toolId: string;
  target: string;
  code: string;
  paramSchema: unknown;
  defaultParams: unknown;
  assetSlots: unknown;
  plan?: unknown;
  createdAt: string;
};

export type ToolResponse = {
  id: string;
  publicId: string;
  ownerUserId: string;
  status: string;
  title?: string | null;
  description?: string | null;
  /** M8a gallery tags */
  tags?: string[];
  publishedAt?: string | null;
  publishedVersionId?: string | null;
  /** M8b: eligible for gallery list after gates */
  galleryReady?: boolean;
  exportSmokeAt?: string | null;
  /** M8c gallery thumbnail */
  thumbnailAssetId?: string | null;
  thumbnailUrl?: string | null;
  createdAt: string;
  updatedAt: string;
  latestVersion?: ToolVersionResponse | null;
  /** M5c: owner Studio param overlay */
  draftParams?: Record<string, unknown>;
  /** M5c: slotId → http URL | null */
  draftAssets?: Record<string, string | null>;
  /** Continuous Studio refine chat (tool-scoped) */
  chatHistory?: Array<{
    id?: string;
    role: string;
    content: string;
    kind?: string;
    createdAt?: string;
    meta?: Record<string, unknown>;
  }>;
};

/** Structured gate failure from gallery publish 422. */
export type PublishGateFailure = {
  code: string;
  message: string;
};

/** M8a/M8b optional body for POST /publish (thin share = {}). */
export type ToolPublishRequest = {
  title?: string | null;
  description?: string | null;
  tags?: string[];
  /** When true, snap draftParams into a new version's defaultParams. */
  freezeDraft?: boolean;
  /** When true, run quality gates and set galleryReady. */
  forGallery?: boolean;
  /** Client proved captureFrame / PNG export (required when forGallery). */
  exportSmokeOk?: boolean;
  /** M8c: asset id from upload kind=thumb (required for gallery). */
  thumbnailAssetId?: string | null;
};

export type ToolDraftPatch = {
  draftParams?: Record<string, unknown>;
  draftAssets?: Record<string, string | null>;
};

/** M7d — anonymous public tool payload (no owner / draft fields). */
export type PublicToolVersionResponse = {
  id: string;
  target: string;
  code: string;
  paramSchema: unknown;
  defaultParams: unknown;
  assetSlots: unknown;
};

export type PublicToolResponse = {
  publicId: string;
  status: string;
  title?: string | null;
  description?: string | null;
  tags?: string[];
  publishedAt?: string | null;
  publishedVersionId?: string | null;
  thumbnailAssetId?: string | null;
  /** CORS-safe raw URL for gallery cards / <img> */
  thumbnailUrl?: string | null;
  version: PublicToolVersionResponse;
};

export type OwnerToolKind = "all" | "created" | "remixed";

/** Thin owner library card from GET /api/v1/tools — no code / draft / chat. */
export type OwnerToolCard = {
  id: string;
  publicId: string;
  title?: string | null;
  status: string;
  galleryReady?: boolean;
  thumbnailUrl?: string | null;
  updatedAt: string;
  publishedAt?: string | null;
  isRemix: boolean;
  hasRunnableVersion: boolean;
  tags?: string[];
};

export type OwnerToolListResponse = {
  items: OwnerToolCard[];
  limit: number;
  offset: number;
  hasMore: boolean;
};

export type ListMyToolsOptions = {
  limit?: number;
  offset?: number;
  kind?: OwnerToolKind;
};

/**
 * GET /api/v1/tools — signed-in owner's library (profile).
 */
export async function listMyTools(
  options?: ListMyToolsOptions,
): Promise<OwnerToolListResponse> {
  const params = new URLSearchParams();
  if (options?.limit != null) params.set("limit", String(options.limit));
  if (options?.offset != null) params.set("offset", String(options.offset));
  if (options?.kind && options.kind !== "all") params.set("kind", options.kind);
  const qs = params.toString();
  const url = `${getApiBaseUrl()}/api/v1/tools${qs ? `?${qs}` : ""}`;

  const res = await fetch(url, {
    method: "GET",
    credentials: "include",
  });

  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(
      `List tools failed (${res.status})${text ? `: ${text}` : ""}`,
    );
  }

  return res.json() as Promise<OwnerToolListResponse>;
}

/** Structured error from POST /tools/fork/{publicId}. */
export class ForkToolError extends Error {
  readonly status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = "ForkToolError";
    this.status = status;
  }
}

/**
 * POST /api/v1/tools/fork/{publicId} — clone a published tool as a draft.
 * Caller becomes the owner. 401/404/other preserved on ForkToolError.status.
 */
export async function forkTool(publicId: string): Promise<ToolResponse> {
  const res = await fetch(
    `${getApiBaseUrl()}/api/v1/tools/fork/${encodeURIComponent(publicId)}`,
    {
      method: "POST",
      credentials: "include",
    },
  );

  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new ForkToolError(
      `Fork tool failed (${res.status})${text ? `: ${text}` : ""}`,
      res.status,
    );
  }

  return res.json() as Promise<ToolResponse>;
}

/** GET /api/v1/tools/{toolId} — owner only. */
export async function getTool(toolId: string): Promise<ToolResponse> {
  const res = await fetch(
    `${getApiBaseUrl()}/api/v1/tools/${encodeURIComponent(toolId)}`,
    {
      method: "GET",
      credentials: "include",
    },
  );

  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(
      `Get tool failed (${res.status})${text ? `: ${text}` : ""}`,
    );
  }

  return res.json() as Promise<ToolResponse>;
}

/**
 * PATCH /api/v1/tools/{toolId}/draft — owner only (M5c).
 * Full-replace for bags present; at least one bag required.
 */
export async function patchToolDraft(
  toolId: string,
  patch: ToolDraftPatch,
): Promise<ToolResponse> {
  if (patch.draftParams === undefined && patch.draftAssets === undefined) {
    throw new Error("patchToolDraft requires draftParams and/or draftAssets");
  }

  const res = await fetch(
    `${getApiBaseUrl()}/api/v1/tools/${encodeURIComponent(toolId)}/draft`,
    {
      method: "PATCH",
      credentials: "include",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        draftParams: patch.draftParams,
        draftAssets: patch.draftAssets,
      }),
    },
  );

  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(
      `Patch tool draft failed (${res.status})${text ? `: ${text}` : ""}`,
    );
  }

  return res.json() as Promise<ToolResponse>;
}

/**
 * GET /api/v1/public/tools/{publicId} — no auth (M7d).
 * 404 when missing or still draft.
 */
export async function getPublicTool(
  publicId: string,
): Promise<PublicToolResponse> {
  const res = await fetch(
    `${getApiBaseUrl()}/api/v1/public/tools/${encodeURIComponent(publicId)}`,
    {
      method: "GET",
      credentials: "omit",
    },
  );

  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(
      `Get public tool failed (${res.status})${text ? `: ${text}` : ""}`,
    );
  }

  return res.json() as Promise<PublicToolResponse>;
}

export class PublishGatesError extends Error {
  readonly status: number;
  readonly code: string;
  readonly gates: PublishGateFailure[];

  constructor(
    message: string,
    opts: { status: number; code: string; gates: PublishGateFailure[] },
  ) {
    super(message);
    this.name = "PublishGatesError";
    this.status = opts.status;
    this.code = opts.code;
    this.gates = opts.gates;
  }
}

/**
 * POST /api/v1/tools/{toolId}/publish — owner publish (M7 thin + M8a/M8b).
 * Sets status=published, freezes publishedVersionId.
 * forGallery + exportSmokeOk → galleryReady after gates; else thin share only.
 */
export async function publishTool(
  toolId: string,
  body?: ToolPublishRequest,
): Promise<ToolResponse> {
  const res = await fetch(
    `${getApiBaseUrl()}/api/v1/tools/${encodeURIComponent(toolId)}/publish`,
    {
      method: "POST",
      credentials: "include",
      headers: body ? { "Content-Type": "application/json" } : undefined,
      body: body ? JSON.stringify(body) : undefined,
    },
  );

  if (!res.ok) {
    const text = await res.text().catch(() => "");
    if (res.status === 422 && text) {
      try {
        const parsed = JSON.parse(text) as {
          detail?: {
            code?: string;
            message?: string;
            gates?: PublishGateFailure[];
          };
        };
        const detail = parsed.detail;
        if (
          detail &&
          detail.code === "GATES_FAILED" &&
          Array.isArray(detail.gates)
        ) {
          throw new PublishGatesError(
            detail.message || "Publish gates failed",
            {
              status: res.status,
              code: detail.code,
              gates: detail.gates,
            },
          );
        }
      } catch (e) {
        if (e instanceof PublishGatesError) throw e;
      }
    }
    throw new Error(
      `Publish tool failed (${res.status})${text ? `: ${text}` : ""}`,
    );
  }

  return res.json() as Promise<ToolResponse>;
}

/**
 * POST /api/v1/tools/{toolId}/unpublish — owner full takedown (M8f).
 * status=draft, galleryReady=false; public /t and gallery hide the tool.
 */
export async function unpublishTool(toolId: string): Promise<ToolResponse> {
  const res = await fetch(
    `${getApiBaseUrl()}/api/v1/tools/${encodeURIComponent(toolId)}/unpublish`,
    {
      method: "POST",
      credentials: "include",
    },
  );

  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(
      `Unpublish tool failed (${res.status})${text ? `: ${text}` : ""}`,
    );
  }

  return res.json() as Promise<ToolResponse>;
}
